src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +150,143 @@ private void logTargetPath() {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
-        final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
-        if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
-            return null;
+    private void addVisionObservation(Pose2d estimatedPose, StandardDeviations standardDeviations, double timestamp) {
+        if (isVisionObservationTooOld(timestamp))
+            return;
+
+        final Pose2d estimatedPoseAtTimestamp = getPoseAtTimestamp(timestamp);
+        if (estimatedPoseAtTimestamp == null)
+            return;
+
+        final Pose2d estimatedPoseWithAmbiguityCompensation = calculateEstimatedPoseWithAmbiguityCompensation(estimatedPoseAtTimestamp, estimatedPose, standardDeviations);
+        final Transform2d estimatedPoseAtTimestampToOdometryPose = new Transform2d(estimatedPoseAtTimestamp, odometryPose);
 
-        return new PoseEstimator6328.VisionObservation(
-                robotPose,
-                aprilTagCamera.getLatestResultTimestampSeconds(),
-                aprilTagCamera.calculateStandardDeviations()
-        );
+        this.estimatedPose = estimatedPoseWithAmbiguityCompensation.plus(estimatedPoseAtTimestampToOdometryPose);
     }
 
-    private void putAprilTagsOnFieldWidget() {
-        for (Map.Entry<Integer, Pose3d> entry : FieldConstants.TAG_ID_TO_POSE.entrySet()) {
-            final Pose2d tagPose = entry.getValue().toPose2d();
-            field.getObject("Tag " + entry.getKey()).setPose(tagPose);
+    /**
+     * Checks if the buffer that stores the previous poses of the robot still has something stored at the target timestamp.
+     *
+     * @param timestamp the target timestamp to check
+     * @return whether the buffer contains a value at that timestamp or not
+     */
+    private boolean isVisionObservationTooOld(double timestamp) {
+        try {
+            final double oldestEstimatedRobotPoseTimestamp = previousOdometryPoses.getInternalBuffer().lastKey() - PoseEstimatorConstants.POSE_BUFFER_SIZE_SECONDS;
+            if (oldestEstimatedRobotPoseTimestamp > timestamp)
+                return true;
+        } catch (NoSuchElementException e) {
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Calculates the estimated pose of a vision observation with compensation for its ambiguity.
+     *
+     * @param estimatedPoseAtTimestamp      the estimated pose of the robot at the time of the observation
+     * @param observationEstimatedPose      the estimated robot pose from the observation
+     * @param observationStandardDeviations the ambiguity of the observation
+     * @return the estimated pose with compensation for its ambiguity
+     */
+    private Pose2d calculateEstimatedPoseWithAmbiguityCompensation(Pose2d estimatedPoseAtTimestamp, Pose2d observationEstimatedPose, StandardDeviations observationStandardDeviations) {
```

Comment:
This name doesn't sit right with me

Reply:
You got any better ideas?

Reply:
I suppose it's alright, as long as you explain it very well in the jdoc, as well as define ambiguity compensation. Same for calculateAllowedMovementFromAmbiguity